### ISSUE TITLE
Potential bug in PPO Trainer

### ISSUE THREAD

**Issue by kyleliang919** [huggingface/trl#2596]
### Reproduction

```bash
accelerate launch examples/scripts/ppo/ppo_tldr.py \
    --dataset_name trl-internal-testing/tldr-preference-sft-trl-style \
    --dataset_test_split validation \
    --output_dir model/minimal/ppo_tldr \
    --learning_rate 3e-6 \
    --per_device_train_batch_size 16 \
    --gradient_accumulation_steps 32 \
    --total_episodes 1000000 \
    --model_name_or_path EleutherAI/pythia-1b-deduped \
    --sft_model_path cleanrl/EleutherAI_pythia-1b-deduped__sft__tldr \
    --reward_model_path cleanrl/EleutherAI_pythia-1b-deduped__reward__tldr \
    --local_rollout_forward_batch_size 16 \
    --missing_eos_penalty 1.0 \
    --stop_token eos \
    --eval_strategy steps \
    --eval_steps 100

```

outputs:
Can't replicate the results here:
https://wandb.ai/huggingface/trl/runs/dd2o3g35/overview

Supposedly the same command except that I only ran on a single GPU, due to resource constraint. But I set gradient accumulation steps to 32 to make effective batch size the same in the end. Nothing has been changed beyond that.
Score curves:
![Image](https://github.com/user-attachments/assets/dda88732-4f2e-4b8d-a9a9-2a830819791e)

### System Info

- Platform: Linux-5.14.0-362.24.1.el9_3.aarch64+64k-aarch64-with-glibc2.34
- Python version: 3.9.18
- PyTorch version: 2.5.1
- CUDA device(s): NVIDIA GH200 120GB
- Transformers version: 4.47.1
- Accelerate version: 1.2.1
- Accelerate config:
  - compute_environment: LOCAL_MACHINE
  - distributed_type: NO
  - mixed_precision: bf16
  - use_cpu: False
  - debug: False
  - num_processes: 1
  - machine_rank: 0
  - num_machines: 1
  - gpu_ids: 0
  - rdzv_backend: static
  - same_network: True
  - main_training_function: main
  - enable_cpu_affinity: True
  - downcast_bf16: no
  - tpu_use_cluster: False
  - tpu_use_sudo: False
  - tpu_env: []
  - dynamo_config: {'dynamo_backend': 'EAGER'}
- Datasets version: 3.2.0
- HF Hub version: 0.27.0
- TRL version: 0.13.0
- bitsandbytes version: not installed
- DeepSpeed version: 0.16.2
- Diffusers version: not installed
- Liger-Kernel version: not installed
- LLM-Blender version: not installed
- OpenAI version: not installed
- PEFT version: 0.14.0

### Checklist

- [x] I have checked that my issue isn't already filed (see [open issues](https://github.com/huggingface/trl/issues?q=is%3Aissue))
- [x] I have included my system information
- [x] Any code provided is minimal, complete, and reproducible ([more on MREs](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks))
- [x] Any code provided is properly formatted in code blocks, (no screenshot, [more on code blocks](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks))
- [x] Any traceback provided is complete

**Comment by superkevingit** [comment]
Just wanted to follow up to see if there is any solution for this issue.

**Comment by qgallouedec** [comment]
Hey @kyleliang919 (cc @superkevingit), sorry for the slow reply. This is expected rather than a bug.

PPO whitens advantages per-rank, over the local rollout batch (per_device_train_batch_size × gradient_accumulation_steps), with no cross-rank gather https://github.com/huggingface/trl/blob/ceae827cb3a772efb4b98ede070a46e15aee7265/trl/experimental/ppo/ppo_trainer.py#L797

So:
grad-accum vs. per-device batch (same GPU count): invariant, the product local_batch_size is unchanged, so the whitening population is identical.
number of GPUs (fixed effective batch): not invariant. Your 1-GPU run whitens over one large local batch (~512), while the 8-GPU reference whitens over each rank's smaller batch independently; 8 mean/std estimates instead of 1. Different stats → different curves.

So the mismatch comes from the change in data-parallel degree, not grad-accum. The per-rank behavior is inherited from the reference PPO implementation. It's arguably worth making global (a cross-rank reduction of mean/std would be cheap), but that's an enhancement, not a bug fix; so closing as not-a-bug.

Final thing, this enhancement is not open to contributions, PPO is now in maintenance mode, so we won't modify it anymore.
